index.js

```diff
@@ -123,15 +133,102 @@ Ycb.prototype = {
                 n = keys.length;
                 for(j=0; j<n; j++) {
-                    if(cur[keys[j]] !== undefined && seen[keys[j]] === undefined) {
-                        this._readHelper(cur[keys[j]], depth+1, context, collector, subKey);
+                    if(cur.has(keys[j]) && seen[keys[j]] === undefined) {
+                        this._readHelper(cur.get(keys[j]), depth+1, context, collector, subKey);
                         seen[keys[j]] = true;
                     }
                 }
             }
         }
     },
 
+    /**
+     * Read the configs for the given context and time and merge them in order general to specific.
+     * Scheduled configs with equivalent contexts are merged by starting time.
+     * @method read
+     * @param contextObj {object}
+     * @param time {number} time in milliseconds
+     * @param options {object}
+     * @return {array}
+     */
+    readTimeAware: function(contextObj, time, options) {
+        options = options ? mergeDeep(this.options, options, true) : this.options;
+        var context = this._parseContext(contextObj);
+        var subKey = options.applySubstitutions !== false ? 'subbed': 'unsubbed';
+        var collector = {};
+        var soonest = SENTINEL_TIME;
+        if(this.masterDelta) {
+            if(this.masterDelta[subKey]) {
+                collector = cloneDeep(this.masterDelta[subKey]);
+            }
+            if(this.masterDelta.schedules) {
+                soonest = this._readScheduled(this.masterDelta.schedules, time, collector, subKey);
+            }
+        }
+        var ret = this._readTimeAwareHelper(this.tree, 0, context, time, collector, subKey);
+        if(ret < soonest) {
+            soonest = ret;
+        }
+        if(soonest !== SENTINEL_TIME && options.cacheInfo === true) {
+            collector[EXPIRATION_KEY] = soonest;
+        }
+        return collector;
+    },
+
+    /**
+     * Recurse through the tree merging configs that apply to the given context.
+     * @param cur {object} the current node.
+     * @param depth {number} current depth in tree.
+     * @param context {array} the context in internal format.
+     * @param time {number} milliseconds
+     * @param collector {object} the config we build
+     * @param subKey {string} determines if substituted or non-substituted configs are used.
+     * @private
+     */
+    _readTimeAwareHelper: function(cur, depth, context, time, collector, subKey) {
+        var soonest = SENTINEL_TIME;
+        var ret;
+        if(depth === context.length) {
+            if(cur[subKey]) {
+                mergeDeep(cur[subKey], collector, false);
+            }
+            if(cur.schedules) {
+                soonest = this._readScheduled(cur.schedules, time, collector, subKey);
+            }
+            return soonest;
+        }
+        var value = context[depth];
+        if(!Array.isArray(value)) {
+            var keys = this.precedenceMap[value];
```

Comment:
Can you consolidate this code with code from lines 213-228? Looks like they do the same thing.